## Задание №12

Необходимо восстановить поле, которое образуется в результате продувки в аэродинамической трубе.

**Формат ввода**

Вам задан набор данных (`train.csv`), состоящий из следующих полей:

- $x$, $y$, $z$ — координаты точки в пространстве;
- $u_x$, $u_y$, $u_z$ — вектор скорости потока в данной точке;
- $p$ — давление в данной точке.

Набор данных `test.csv` содержит только координаты $(x, y, z)$.

**Формат вывода**

Вам необходимо послать в тестирующую систему файл `submission.csv`, $k$-я строка которого будет содержать предсказанные значения $u_x$, $u_y$, $u_z$, $p$ для $k$-ой строки файла `test.csv`. Предсказанные таргеты в каждой строке разделяются запятыми. Нулевая строка должна содержать название колонок (игнорируются при чтении). Файл `sample_submission.csv` представляет из себя пример посылки.

**Примечания**

Оценка решения будет проводится на основе метрики MAE по формуле:

$$points = \min\left(10, \max\left(0, \frac{MaxScore - Score}{MaxScore - MinScore}\right)\right)$$

где $Score$ — значение MAE вашей посылки, $MinScore$ — порог по метрике снизу, $MaxScore$ — порог по метрике сверху.

Для различных полей значения $MinScore$ и $MaxScore$ различаются:

- Для $u_x$, $u_y$, $u_z$ значения $MinScore = 1.1$ и $MaxScore = 1.9$;
- Для $p$ значения $MinScore = 500$ и $MaxScore = 1900$.

После подсчёта $points$ по каждому таргету по-отдельности итоговые значения $points$ усредняются по всем четырём таргетам.

**Корректировка:**
Максимальный балл при `MAE` < 0.13

In [12]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
%matplotlib inline

import os
import tensorflow as tf
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

from tensorflow.keras.utils import plot_model
from tensorflow.keras import layers, losses, metrics, regularizers, Sequential
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint


sns.set()

train_path = 'train.csv'
test_path = 'test.csv'
sample_path = 'sample_submission.csv'
submission_path = 'submission.csv'

In [13]:
train_csv = pd.read_csv(train_path)
test_csv = pd.read_csv(test_path)
train_csv

x         y         z         u_x        u_y       u_z  \
0    -2.182809 -0.029668 -4.177883  410.436506   0.013901 -0.054928   
1    -5.697199 -1.168055  2.704167  410.437352  -0.030757  0.021365   
2    -2.388080  0.489320 -4.422072  410.443844   0.028441 -0.042093   
3    -6.859628  0.177311 -1.745587  410.430337   0.008494 -0.011427   
4     2.668538 -0.702350  3.033114  417.831936  -0.717926  0.116151   
...        ...       ...       ...         ...        ...       ...   
8527 -4.240097 -0.190458 -0.233550  406.192347  19.419476 -7.903454   
8528 -2.490715 -0.443627 -3.005777  410.190644  -0.012390 -0.552977   
8529  1.819966 -0.805841 -4.031370  409.337590  -0.234673 -1.674802   
8530 -1.464362 -1.064565  3.252417  410.305496  -0.110896  0.647835   
8531  1.819498 -0.691653  0.839134  442.477173  16.781423 -3.749191   

                  p  
0     100030.436265  
1     100020.971632  
2     100021.156537  
3     100018.526121  
4      97710.352064  
...             ...  
8527  102458.496025  
8528  100225.470127  
8529  100720.371138  
8530  100233.838109  
8531   82108.472492  

[8532 rows x 7 columns]

In [14]:
test_csv

x         y         z
0     12.595389 -0.676479  9.448017
1     -5.115985  0.332546 -1.952567
2     -2.182809 -0.133158  4.463463
3      2.233563 -0.857583 -3.054615
4     -0.847552 -0.185422 -1.177141
...         ...       ...       ...
3651  -2.695986 -0.598862 -3.103453
3652  12.134517 -0.883459 -7.321329
3653  -3.927776 -1.165944  0.220514
3654   1.306790 -0.909331  3.745842
3655   2.668540  0.436024 -3.005773

[3656 rows x 3 columns]

In [15]:
feat_cols = ['x', 'y', 'z']
target_cols = ['u_x', 'u_y', 'u_z', 'p']

n_feats = len(feat_cols)
n_targets = len(target_cols)

In [16]:
def generate_train_data():
    mean, std = train_csv[target_cols].mean(), train_csv[target_cols].std()
    for (i, data) in train_csv.iterrows():
        yield (data[feat_cols], (data[target_cols] - mean) / std)


def prep_train_data():
    return tf.data.Dataset.from_generator(
        generate_train_data, 
        output_signature=(
            tf.TensorSpec(shape=(n_feats), dtype=tf.float32), 
            tf.TensorSpec(shape=(n_targets), dtype=tf.float32)
        )
    )

In [17]:
def prep_linear(inputs, outputs, layer_sizes, use_dropout=False, use_regularizers=False):
    model = Sequential()
    model.add(layers.InputLayer(input_shape=(inputs)))
    
    for layer_size in layer_sizes:
        if use_regularizers:
            model.add(layers.Dense(layer_size, kernel_regularizer=regularizers.L2(1e-4)))
        else:
            model.add(layers.Dense(layer_size))
        model.add(layers.PReLU()) # PReLU > LeakyReLU
        if use_dropout:
            model.add(layers.Dropout(.25))
    
    model.add(layers.Dense(outputs))
    return model


def prep_unet(inputs, outputs, layer_sizes):
    inputs = tf.keras.Input((inputs))
    x = inputs

    skips = []

    for layer_size in layer_sizes:
        x = layers.Dense(layer_size)(x)
        x = layers.PReLU()(x) # PReLU > LeakyReLU
        skips.append(x)

    for layer_size, skip in zip(layer_sizes[-2::-1], skips[-2::-1]):
        x = layers.Dense(layer_size)(x)
        x = layers.PReLU()(x) # PReLU > LeakyReLU
        x = tf.keras.layers.concatenate([x, skip])

    model = tf.keras.Model(inputs, layers.Dense(outputs)(x))
    return model

In [18]:
def build_regression(model):
    model.compile(
        optimizer='adam', 
        loss=losses.MeanAbsoluteError(),
        metrics=[metrics.RootMeanSquaredError(name='rmse')]
    )


def prep_simple():
    dense_model = prep_linear(n_feats, n_targets, [32 * n_feats, 64 * n_feats, 16 * n_feats], use_dropout=True)
    build_regression(dense_model)
    return dense_model
    

def prep_parallel():
    inputs = tf.keras.Input((n_feats))
    
    speed_model = prep_linear(n_feats, n_targets - 1, [32 * n_feats, 64 * n_feats, 16 * n_feats], use_dropout=True, use_regularizers=True)
    pressure_model = prep_linear(n_feats, 1, [16 * n_feats, 32 * n_feats, 8 * n_feats], use_dropout=True, use_regularizers=False)

    outputs = layers.concatenate([speed_model(inputs), pressure_model(inputs)])
    parallel_model = tf.keras.Model(inputs, outputs)

    build_regression(parallel_model)
    
    return parallel_model


def prep_parallel_unet():
    inputs = tf.keras.Input((n_feats))
    
    speed_model = prep_unet(n_feats, n_targets - 1, [8 * n_feats, 16 * n_feats, 32 * n_feats])
    pressure_model = prep_unet(n_feats, 1, [4 * n_feats, 8 * n_feats, 16 * n_feats])

    outputs = layers.concatenate([speed_model(inputs), pressure_model(inputs)])
    parallel_model = tf.keras.Model(inputs, outputs)

    build_regression(parallel_model)
    
    return parallel_model

In [19]:
model = prep_parallel_unet()
plot_model(model, '12.png', show_shapes=True)
model.summary()

You must install pydot (`pip install pydot`) and install graphviz (see instructions at https://graphviz.gitlab.io/download/) for plot_model to work.
Model: "model_5"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_4 (InputLayer)        [(None, 3)]                  0         []                            
                                                                                                  
 model_3 (Functional)        (None, 3)                    13371     ['input_4[0][0]']             
                                                                                                  
 model_4 (Functional)        (None, 1)                    3469      ['input_4[0][0]']             
                                                                                                  
 concatenate_9 (Concatenate  (None, 4)    

In [20]:
BATCH_SIZE = 64
BUFFER_SIZE = 2000
VAL_SIZE = len(train_csv.index) // 7

ds = prep_train_data()

train_ds = ds.skip(VAL_SIZE).shuffle(BUFFER_SIZE).batch(BATCH_SIZE).prefetch(tf.data.AUTOTUNE)
val_ds = ds.take(VAL_SIZE).batch(BATCH_SIZE).prefetch(tf.data.AUTOTUNE)

In [21]:
training_folder = os.path.join('dnn', '12')
tensoboard_path = os.path.join(training_folder, 'logs_cpu')
model_path = os.path.join(training_folder, 'best_cpu')

model_callbacks = [
    EarlyStopping(patience=20),
    ModelCheckpoint(model_path, monitor='val_loss', save_best_only=True, mode='min')
]

In [22]:
model.fit(train_ds, validation_data=val_ds, epochs=200, callbacks=model_callbacks)

Epoch 1/200


     83/Unknown - 8s 48ms/step - loss: 0.4859 - rmse: 1.0628INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 11s 61ms/step - loss: 0.4653 - rmse: 1.0418 - val_loss: 0.3885 - val_rmse: 0.9300
Epoch 2/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.4213 - rmse: 1.0216INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.4069 - rmse: 0.9983 - val_loss: 0.3786 - val_rmse: 0.9208
Epoch 3/200
115/115 [==============================] - 7s 42ms/step - loss: 0.4000 - rmse: 0.9909 - val_loss: 0.3831 - val_rmse: 0.9156
Epoch 4/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.4008 - rmse: 1.0084INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.3878 - rmse: 0.9782 - val_loss: 0.3611 - val_rmse: 0.8954
Epoch 5/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.3734 - rmse: 0.9674INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.3692 - rmse: 0.9568 - val_loss: 0.3371 - val_rmse: 0.8646
Epoch 6/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.3501 - rmse: 0.9348INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.3495 - rmse: 0.9270 - val_loss: 0.3236 - val_rmse: 0.8369
Epoch 7/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.3316 - rmse: 0.9070INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 55ms/step - loss: 0.3318 - rmse: 0.8983 - val_loss: 0.3087 - val_rmse: 0.8064
Epoch 8/200
111/115 [===========================>..] - ETA: 0s - loss: 0.3153 - rmse: 0.8746INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 60ms/step - loss: 0.3139 - rmse: 0.8693 - val_loss: 0.2887 - val_rmse: 0.7708
Epoch 9/200
112/115 [============================>.] - ETA: 0s - loss: 0.2924 - rmse: 0.8328INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 56ms/step - loss: 0.2949 - rmse: 0.8419 - val_loss: 0.2738 - val_rmse: 0.7429
Epoch 10/200
113/115 [============================>.] - ETA: 0s - loss: 0.2811 - rmse: 0.8161INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.2812 - rmse: 0.8156 - val_loss: 0.2628 - val_rmse: 0.7207
Epoch 11/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.2685 - rmse: 0.7964INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 56ms/step - loss: 0.2690 - rmse: 0.7890 - val_loss: 0.2452 - val_rmse: 0.6849
Epoch 12/200
115/115 [==============================] - ETA: 0s - loss: 0.2574 - rmse: 0.7608INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.2574 - rmse: 0.7608 - val_loss: 0.2382 - val_rmse: 0.6641
Epoch 13/200
112/115 [============================>.] - ETA: 0s - loss: 0.2466 - rmse: 0.7435INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 61ms/step - loss: 0.2453 - rmse: 0.7397 - val_loss: 0.2317 - val_rmse: 0.6495
Epoch 14/200
114/115 [============================>.] - ETA: 0s - loss: 0.2367 - rmse: 0.7228INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.2367 - rmse: 0.7226 - val_loss: 0.2184 - val_rmse: 0.6266
Epoch 15/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.2328 - rmse: 0.7296INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.2273 - rmse: 0.7086 - val_loss: 0.2166 - val_rmse: 0.6255
Epoch 16/200
 85/115 [=====================>........] - ETA: 1s - loss: 0.2243 - rmse: 0.6984INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.2220 - rmse: 0.6951 - val_loss: 0.2097 - val_rmse: 0.6096
Epoch 17/200
115/115 [==============================] - ETA: 0s - loss: 0.2187 - rmse: 0.6869INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.2187 - rmse: 0.6869 - val_loss: 0.2022 - val_rmse: 0.5960
Epoch 18/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.2141 - rmse: 0.6988INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.2110 - rmse: 0.6721 - val_loss: 0.1999 - val_rmse: 0.5944
Epoch 19/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.2127 - rmse: 0.6823INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.2084 - rmse: 0.6587 - val_loss: 0.1961 - val_rmse: 0.5762
Epoch 20/200
111/115 [===========================>..] - ETA: 0s - loss: 0.2014 - rmse: 0.6459INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.2012 - rmse: 0.6451 - val_loss: 0.1956 - val_rmse: 0.5586
Epoch 21/200
112/115 [============================>.] - ETA: 0s - loss: 0.1956 - rmse: 0.6295INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 10s 63ms/step - loss: 0.1959 - rmse: 0.6292 - val_loss: 0.1869 - val_rmse: 0.5507
Epoch 22/200
114/115 [============================>.] - ETA: 0s - loss: 0.1915 - rmse: 0.6138INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 60ms/step - loss: 0.1911 - rmse: 0.6130 - val_loss: 0.1766 - val_rmse: 0.5283
Epoch 23/200
114/115 [============================>.] - ETA: 0s - loss: 0.1892 - rmse: 0.6002INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 61ms/step - loss: 0.1891 - rmse: 0.5998 - val_loss: 0.1753 - val_rmse: 0.5158
Epoch 24/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.1858 - rmse: 0.5979INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 61ms/step - loss: 0.1837 - rmse: 0.5890 - val_loss: 0.1736 - val_rmse: 0.5046
Epoch 25/200
115/115 [==============================] - ETA: 0s - loss: 0.1791 - rmse: 0.5740INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.1791 - rmse: 0.5740 - val_loss: 0.1705 - val_rmse: 0.5013
Epoch 26/200
114/115 [============================>.] - ETA: 0s - loss: 0.1778 - rmse: 0.5665INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 10s 62ms/step - loss: 0.1776 - rmse: 0.5660 - val_loss: 0.1678 - val_rmse: 0.4929
Epoch 27/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.1790 - rmse: 0.5818INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1736 - rmse: 0.5561 - val_loss: 0.1630 - val_rmse: 0.4815
Epoch 28/200
114/115 [============================>.] - ETA: 0s - loss: 0.1693 - rmse: 0.5478INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1694 - rmse: 0.5475 - val_loss: 0.1615 - val_rmse: 0.4759
Epoch 29/200
 86/115 [=====================>........] - ETA: 1s - loss: 0.1746 - rmse: 0.5747INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.1648 - rmse: 0.5364 - val_loss: 0.1601 - val_rmse: 0.4711
Epoch 30/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1623 - rmse: 0.5307 - val_loss: 0.1604 - val_rmse: 0.4662
Epoch 31/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1601 - rmse: 0.5205 - val_loss: 0.1658 - val_rmse: 0.4697
Epoch 32/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1603 - rmse: 0.5202INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1577 - rmse: 0.5150 - val_loss: 0.1560 - val_rmse: 0.4568
Epoch 33/200
115/115 [==============================] - 8s 45ms/step - loss: 0.1565 - rmse: 0.5113 - val_loss: 0.1637 - val_rmse: 0.4631
Epoch 34/200
114/115 [============================>.] - ETA: 0s - loss: 0.1556 - rmse: 0.5000INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 61ms/step - loss: 0.1556 - rmse: 0.4996 - val_loss: 0.1500 - val_rmse: 0.4422
Epoch 35/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1541 - rmse: 0.4965 - val_loss: 0.1507 - val_rmse: 0.4395
Epoch 36/200
115/115 [==============================] - 8s 45ms/step - loss: 0.1502 - rmse: 0.4919 - val_loss: 0.1561 - val_rmse: 0.4412
Epoch 37/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.1494 - rmse: 0.4906INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1514 - rmse: 0.4896 - val_loss: 0.1495 - val_rmse: 0.4316
Epoch 38/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1492 - rmse: 0.4978INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1473 - rmse: 0.4808 - val_loss: 0.1466 - val_rmse: 0.4296
Epoch 39/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1447 - rmse: 0.4759 - val_loss: 0.1473 - val_rmse: 0.4237
Epoch 40/200
112/115 [============================>.] - ETA: 0s - loss: 0.1456 - rmse: 0.4704INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 62ms/step - loss: 0.1450 - rmse: 0.4702 - val_loss: 0.1453 - val_rmse: 0.4169
Epoch 41/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1423 - rmse: 0.4637 - val_loss: 0.1481 - val_rmse: 0.4216
Epoch 42/200
110/115 [===========================>..] - ETA: 0s - loss: 0.1421 - rmse: 0.4663INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1421 - rmse: 0.4628 - val_loss: 0.1392 - val_rmse: 0.4083
Epoch 43/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1384 - rmse: 0.4560 - val_loss: 0.1423 - val_rmse: 0.4102
Epoch 44/200
114/115 [============================>.] - ETA: 0s - loss: 0.1378 - rmse: 0.4528INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 60ms/step - loss: 0.1383 - rmse: 0.4559 - val_loss: 0.1373 - val_rmse: 0.4066
Epoch 45/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1360 - rmse: 0.4485 - val_loss: 0.1427 - val_rmse: 0.4092
Epoch 46/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1353 - rmse: 0.4485 - val_loss: 0.1397 - val_rmse: 0.4038
Epoch 47/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1382 - rmse: 0.4429 - val_loss: 0.1379 - val_rmse: 0.3987
Epoch 48/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1349 - rmse: 0.4454INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1329 - rmse: 0.4397 - val_loss: 0.1358 - val_rmse: 0.3985
Epoch 49/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1369 - rmse: 0.4647INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1321 - rmse: 0.4347 - val_loss: 0.1340 - val_rmse: 0.3952
Epoch 50/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1312 - rmse: 0.4355 - val_loss: 0.1404 - val_rmse: 0.4070
Epoch 51/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1335 - rmse: 0.4386 - val_loss: 0.1375 - val_rmse: 0.3962
Epoch 52/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1310 - rmse: 0.4317 - val_loss: 0.1389 - val_rmse: 0.4045
Epoch 53/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1326 - rmse: 0.4350 - val_loss: 0.1372 - val_rmse: 0.3959
Epoch 54/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1289 - rmse: 0.4258 - val_loss: 0.1388 - val_rmse: 0.3935
Epoch 55/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.1347 - rmse: 0.4575INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1290 - rmse: 0.4247 - val_loss: 0.1267 - val_rmse: 0.3813
Epoch 56/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1282 - rmse: 0.4227 - val_loss: 0.1286 - val_rmse: 0.3821
Epoch 57/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1250 - rmse: 0.4202 - val_loss: 0.1292 - val_rmse: 0.3832
Epoch 58/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1269 - rmse: 0.4228 - val_loss: 0.1318 - val_rmse: 0.3928
Epoch 59/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1249 - rmse: 0.4175 - val_loss: 0.1285 - val_rmse: 0.3810
Epoch 60/200
115/115 [==============================] - 8s 44ms/step - loss: 0.1266 - rmse: 0.4202 - val_loss: 0.1281 - val_rmse: 0.3842
Epoch 61/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1241 - rmse: 0.4164 - val_loss: 0.1336 - val_rmse: 0.3915
Epoch 62/200
115/115 [==============================] 

INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1232 - rmse: 0.4130 - val_loss: 0.1250 - val_rmse: 0.3717
Epoch 64/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1238 - rmse: 0.4098 - val_loss: 0.1272 - val_rmse: 0.3728
Epoch 65/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1207 - rmse: 0.4047 - val_loss: 0.1294 - val_rmse: 0.3917
Epoch 66/200
115/115 [==============================] - 7s 44ms/step - loss: 0.1221 - rmse: 0.4059 - val_loss: 0.1334 - val_rmse: 0.3923
Epoch 67/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1240 - rmse: 0.4147INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 60ms/step - loss: 0.1227 - rmse: 0.4046 - val_loss: 0.1247 - val_rmse: 0.3646
Epoch 68/200
115/115 [==============================] - 7s 44ms/step - loss: 0.1198 - rmse: 0.3997 - val_loss: 0.1257 - val_rmse: 0.3776
Epoch 69/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1252 - rmse: 0.4398INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1185 - rmse: 0.4006 - val_loss: 0.1232 - val_rmse: 0.3757
Epoch 70/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1172 - rmse: 0.3999 - val_loss: 0.1266 - val_rmse: 0.3679
Epoch 71/200
114/115 [============================>.] - ETA: 0s - loss: 0.1185 - rmse: 0.4007INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1184 - rmse: 0.4003 - val_loss: 0.1204 - val_rmse: 0.3651
Epoch 72/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.1201 - rmse: 0.4198INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 60ms/step - loss: 0.1162 - rmse: 0.3957 - val_loss: 0.1201 - val_rmse: 0.3708
Epoch 73/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1162 - rmse: 0.3972 - val_loss: 0.1254 - val_rmse: 0.3674
Epoch 74/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1172 - rmse: 0.3937 - val_loss: 0.1213 - val_rmse: 0.3635
Epoch 75/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1174 - rmse: 0.3901 - val_loss: 0.1252 - val_rmse: 0.3908
Epoch 76/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.1199 - rmse: 0.3930INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1170 - rmse: 0.3946 - val_loss: 0.1163 - val_rmse: 0.3571
Epoch 77/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1155 - rmse: 0.3891 - val_loss: 0.1314 - val_rmse: 0.3880
Epoch 78/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1156 - rmse: 0.3922 - val_loss: 0.1172 - val_rmse: 0.3576
Epoch 79/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1143 - rmse: 0.3860 - val_loss: 0.1214 - val_rmse: 0.3672
Epoch 80/200
115/115 [==============================] - 7s 44ms/step - loss: 0.1167 - rmse: 0.3864 - val_loss: 0.1203 - val_rmse: 0.3647
Epoch 81/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1139 - rmse: 0.3868INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.1138 - rmse: 0.3852 - val_loss: 0.1135 - val_rmse: 0.3586
Epoch 82/200
115/115 [==============================] - 8s 45ms/step - loss: 0.1112 - rmse: 0.3844 - val_loss: 0.1195 - val_rmse: 0.3667
Epoch 83/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1115 - rmse: 0.3820 - val_loss: 0.1161 - val_rmse: 0.3628
Epoch 84/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1145 - rmse: 0.3847 - val_loss: 0.1181 - val_rmse: 0.3526
Epoch 85/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1127 - rmse: 0.3822 - val_loss: 0.1149 - val_rmse: 0.3560
Epoch 86/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1130 - rmse: 0.3840 - val_loss: 0.1144 - val_rmse: 0.3630
Epoch 87/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1105 - rmse: 0.3815 - val_loss: 0.1138 - val_rmse: 0.3596
Epoch 88/200
103/115 [=========================>....] 

INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.1087 - rmse: 0.3740 - val_loss: 0.1099 - val_rmse: 0.3492
Epoch 89/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1087 - rmse: 0.3746 - val_loss: 0.1214 - val_rmse: 0.3711
Epoch 90/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1094 - rmse: 0.3759 - val_loss: 0.1162 - val_rmse: 0.3473
Epoch 91/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1120 - rmse: 0.3801 - val_loss: 0.1163 - val_rmse: 0.3545
Epoch 92/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1080 - rmse: 0.3719 - val_loss: 0.1188 - val_rmse: 0.3816
Epoch 93/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1083 - rmse: 0.3755 - val_loss: 0.1171 - val_rmse: 0.3628
Epoch 94/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1096 - rmse: 0.3738 - val_loss: 0.1146 - val_rmse: 0.3505
Epoch 95/200
115/115 [==============================] 

INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1080 - rmse: 0.3687 - val_loss: 0.1075 - val_rmse: 0.3419
Epoch 97/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1074 - rmse: 0.3727 - val_loss: 0.1082 - val_rmse: 0.3441
Epoch 98/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1086 - rmse: 0.3689 - val_loss: 0.1110 - val_rmse: 0.3433
Epoch 99/200
110/115 [===========================>..] - ETA: 0s - loss: 0.1072 - rmse: 0.3764INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.1064 - rmse: 0.3729 - val_loss: 0.1072 - val_rmse: 0.3375
Epoch 100/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1082 - rmse: 0.3722 - val_loss: 0.1104 - val_rmse: 0.3508
Epoch 101/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1075 - rmse: 0.3690 - val_loss: 0.1083 - val_rmse: 0.3474
Epoch 102/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1049 - rmse: 0.3687INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1058 - rmse: 0.3694 - val_loss: 0.1040 - val_rmse: 0.3363
Epoch 103/200
115/115 [==============================] - 7s 44ms/step - loss: 0.1077 - rmse: 0.3721 - val_loss: 0.1080 - val_rmse: 0.3460
Epoch 104/200
 83/115 [====================>.........] - ETA: 1s - loss: 0.1096 - rmse: 0.3931INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1049 - rmse: 0.3646 - val_loss: 0.1040 - val_rmse: 0.3378
Epoch 105/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1042 - rmse: 0.3642 - val_loss: 0.1078 - val_rmse: 0.3424
Epoch 106/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1055 - rmse: 0.3661 - val_loss: 0.1112 - val_rmse: 0.3494
Epoch 107/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1043 - rmse: 0.3619 - val_loss: 0.1140 - val_rmse: 0.3553
Epoch 108/200
115/115 [==============================] - 7s 45ms/step - loss: 0.1035 - rmse: 0.3614 - val_loss: 0.1050 - val_rmse: 0.3402
Epoch 109/200
114/115 [============================>.] - ETA: 0s - loss: 0.1029 - rmse: 0.3596INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 60ms/step - loss: 0.1028 - rmse: 0.3593 - val_loss: 0.1033 - val_rmse: 0.3381
Epoch 110/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1020 - rmse: 0.3576 - val_loss: 0.1081 - val_rmse: 0.3583
Epoch 111/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1050 - rmse: 0.3617 - val_loss: 0.1038 - val_rmse: 0.3303
Epoch 112/200
115/115 [==============================] - 8s 45ms/step - loss: 0.1044 - rmse: 0.3626 - val_loss: 0.1046 - val_rmse: 0.3479
Epoch 113/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1003 - rmse: 0.3538 - val_loss: 0.1085 - val_rmse: 0.3495
Epoch 114/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1021 - rmse: 0.3582 - val_loss: 0.1121 - val_rmse: 0.3525
Epoch 115/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1039 - rmse: 0.3574 - val_loss: 0.1111 - val_rmse: 0.3459
Epoch 116/200
115/115 [=========================

INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.1032 - rmse: 0.3569 - val_loss: 0.1031 - val_rmse: 0.3384
Epoch 118/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1039 - rmse: 0.3601 - val_loss: 0.1078 - val_rmse: 0.3399
Epoch 119/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1026 - rmse: 0.3552 - val_loss: 0.1148 - val_rmse: 0.3418
Epoch 120/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0994 - rmse: 0.3523 - val_loss: 0.1032 - val_rmse: 0.3323
Epoch 121/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0992 - rmse: 0.3505 - val_loss: 0.1079 - val_rmse: 0.3361
Epoch 122/200
115/115 [==============================] - 7s 42ms/step - loss: 0.1002 - rmse: 0.3530 - val_loss: 0.1074 - val_rmse: 0.3360
Epoch 123/200
115/115 [==============================] - 7s 43ms/step - loss: 0.1026 - rmse: 0.3537 - val_loss: 0.1059 - val_rmse: 0.3379
Epoch 124/200
115/115 [=========================

INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.0984 - rmse: 0.3497 - val_loss: 0.1027 - val_rmse: 0.3379
Epoch 129/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0997 - rmse: 0.3475 - val_loss: 0.1058 - val_rmse: 0.3333
Epoch 130/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0987 - rmse: 0.3516 - val_loss: 0.1051 - val_rmse: 0.3461
Epoch 131/200
115/115 [==============================] - ETA: 0s - loss: 0.0985 - rmse: 0.3476INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.0985 - rmse: 0.3476 - val_loss: 0.1027 - val_rmse: 0.3282
Epoch 132/200
115/115 [==============================] - 7s 44ms/step - loss: 0.0970 - rmse: 0.3425 - val_loss: 0.1094 - val_rmse: 0.3435
Epoch 133/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0985 - rmse: 0.3461 - val_loss: 0.1052 - val_rmse: 0.3322
Epoch 134/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.0974 - rmse: 0.3514INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.0971 - rmse: 0.3443 - val_loss: 0.1016 - val_rmse: 0.3393
Epoch 135/200
115/115 [==============================] - 7s 43ms/step - loss: 0.0974 - rmse: 0.3442 - val_loss: 0.1092 - val_rmse: 0.3566
Epoch 136/200
 84/115 [====================>.........] - ETA: 1s - loss: 0.1005 - rmse: 0.3675INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.0980 - rmse: 0.3481 - val_loss: 0.0991 - val_rmse: 0.3312
Epoch 137/200
113/115 [============================>.] - ETA: 0s - loss: 0.0963 - rmse: 0.3374INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 58ms/step - loss: 0.0964 - rmse: 0.3377 - val_loss: 0.0986 - val_rmse: 0.3396
Epoch 138/200
115/115 [==============================] - 7s 44ms/step - loss: 0.0952 - rmse: 0.3399 - val_loss: 0.1014 - val_rmse: 0.3297
Epoch 139/200
115/115 [==============================] - 7s 44ms/step - loss: 0.0978 - rmse: 0.3436 - val_loss: 0.1091 - val_rmse: 0.3537
Epoch 140/200
115/115 [==============================] - 7s 43ms/step - loss: 0.0970 - rmse: 0.3450 - val_loss: 0.1001 - val_rmse: 0.3378
Epoch 141/200
115/115 [==============================] - 7s 43ms/step - loss: 0.0961 - rmse: 0.3406 - val_loss: 0.1021 - val_rmse: 0.3308
Epoch 142/200
115/115 [==============================] - 7s 44ms/step - loss: 0.0983 - rmse: 0.3425 - val_loss: 0.0990 - val_rmse: 0.3292
Epoch 143/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0949 - rmse: 0.3379 - val_loss: 0.1069 - val_rmse: 0.3366
Epoch 144/200
115/115 [=========================

INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 59ms/step - loss: 0.0958 - rmse: 0.3348 - val_loss: 0.0968 - val_rmse: 0.3237
Epoch 146/200
115/115 [==============================] - 7s 44ms/step - loss: 0.0954 - rmse: 0.3357 - val_loss: 0.0980 - val_rmse: 0.3267
Epoch 147/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0956 - rmse: 0.3399 - val_loss: 0.1034 - val_rmse: 0.3329
Epoch 148/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0945 - rmse: 0.3372 - val_loss: 0.1035 - val_rmse: 0.3380
Epoch 149/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0953 - rmse: 0.3383 - val_loss: 0.1024 - val_rmse: 0.3295
Epoch 150/200
115/115 [==============================] - 7s 44ms/step - loss: 0.0949 - rmse: 0.3377 - val_loss: 0.1086 - val_rmse: 0.3584
Epoch 151/200
115/115 [==============================] - 7s 44ms/step - loss: 0.0934 - rmse: 0.3331 - val_loss: 0.1023 - val_rmse: 0.3214
Epoch 152/200
115/115 [=========================

INFO:tensorflow:Assets written to: dnn\12\best_cpu\assets


115/115 [==============================] - 9s 57ms/step - loss: 0.0896 - rmse: 0.3269 - val_loss: 0.0954 - val_rmse: 0.3158
Epoch 160/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0914 - rmse: 0.3293 - val_loss: 0.0968 - val_rmse: 0.3214
Epoch 161/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0916 - rmse: 0.3284 - val_loss: 0.1068 - val_rmse: 0.3595
Epoch 162/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0952 - rmse: 0.3333 - val_loss: 0.0981 - val_rmse: 0.3199
Epoch 163/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0944 - rmse: 0.3349 - val_loss: 0.1005 - val_rmse: 0.3351
Epoch 164/200
115/115 [==============================] - 8s 44ms/step - loss: 0.0903 - rmse: 0.3261 - val_loss: 0.0968 - val_rmse: 0.3212
Epoch 165/200
115/115 [==============================] - 7s 42ms/step - loss: 0.0932 - rmse: 0.3271 - val_loss: 0.0987 - val_rmse: 0.3211
Epoch 166/200
115/115 [=========================

In [23]:
model = tf.keras.models.load_model(model_path)

In [24]:
train_metrics = model.evaluate(ds.batch(BATCH_SIZE).prefetch(tf.data.AUTOTUNE))

134/134 [==============================] - 7s 48ms/step - loss: 0.0877 - rmse: 0.3193


In [25]:
def generate_test_data():
    for (i, data) in test_csv.iterrows():
        yield data[feat_cols]


def prep_test_data():
    return tf.data.Dataset.from_generator(
        generate_test_data, 
        output_signature=tf.TensorSpec(shape=(n_feats), dtype=tf.float32)
    )

In [26]:
test_ds = prep_test_data().batch(BATCH_SIZE).prefetch(tf.data.AUTOTUNE)
test_outputs = model.predict(test_ds)

mean, std = train_csv[target_cols].mean(), train_csv[target_cols].std()

submission_df = pd.DataFrame(test_outputs, columns=target_cols)
submission_df = submission_df * std + mean
submission_df

58/58 [==============================] - 1s 21ms/step


u_x        u_y       u_z              p
0     410.491489   0.032219  0.335554  100378.855488
1     410.330740   0.101628 -0.070368  100008.853789
2     410.667191  -0.034974  0.044498  100018.734863
3     412.305513  -4.186017 -8.036711  100979.736268
4     410.169659   2.644421 -5.074940  100893.132631
...          ...        ...       ...            ...
3651  410.344329   0.029326 -0.171527  100058.659013
3652  410.377901   0.205342  0.469247  100054.545156
3653  405.690230 -10.395499  1.949544  102946.001675
3654  409.296526   0.072955  1.758138  100174.975668
3655  413.767920   1.640411 -5.484768  100689.935218

[3656 rows x 4 columns]

In [27]:
submission_df.to_csv(submission_path, index=False)

# Результат

* До требуемого результата модель обучилась за **~50** эпох (**0.13** MSE)
* За **~160** эпох модель обучилась до ошибки MSE **0.08**